src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,149 @@
+            }
+
+            /// <summary>
+            /// Finds a matching type such that the display name of the type matches the name passed in, ignoring case. Case isn't used because
+            /// documents with name "Foo.cs" and "foo.cs" should still have the same type name
+            /// </summary>
+            private static async Task<SyntaxNode?> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf(n => !syntaxFacts.IsMethodBody(n)).Where(syntaxFacts.IsTypeDeclaration);
+                return typeDeclarations.FirstOrDefault(d => WorkspacePathUtilities.TypeNameMatchesDocumentName(document, d, syntaxFacts));
+            }
+
+            public static async Task<RenameSymbolDocumentAction?> TryCreateAsync(Document document, string newName, CancellationToken cancellationToken)
+            {
+                var analysis = await AnalyzeAsync(document, newName, cancellationToken).ConfigureAwait(false);
+
+                return analysis.HasValue
+                    ? new RenameSymbolDocumentAction(analysis.Value)
+                    : null;
+            }
+
+            private static async Task<AnalysisResult?> AnalyzeAsync(Document document, string newDocumentName, CancellationToken cancellationToken)
+            {
+                // TODO: Detect naming conflicts ahead of time
+                var documentWithNewName = document.WithName(newDocumentName);
+                var originalSymbolName = Path.GetFileNameWithoutExtension(document.Name);
+                var newTypeName = WorkspacePathUtilities.GetTypeNameFromDocumentName(documentWithNewName);
+
+                if (newTypeName is null)
+                {
+                    return null;
+                }
+
+                var matchingDeclaration = await GetMatchingTypeDeclarationAsync(document, originalSymbolName, cancellationToken).ConfigureAwait(false);
+
+                if (matchingDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingDeclaration, cancellationToken);
+
+                    if (symbol is null || WorkspacePathUtilities.TypeNameMatchesDocumentName(documentWithNewName, symbol.Name))
+                    {
+                        return null;
+                    }
+
+                    return new AnalysisResult(
+                        document,
+                        newDocumentName,
+                        newTypeName,
+                        symbol.Name);
+                }
+
+                return null;
```

Comment:
the inconsistent retursn here bother me.  consider flipping this all to `if (matchingDeclaration is null) return null;`